runtime-modules/bounty/src/lib.rs

```diff
@@ -0,0 +1,619 @@
+
+use frame_support::dispatch::{DispatchError, DispatchResult};
+use frame_support::traits::Currency;
+use frame_support::weights::Weight;
+use frame_support::{decl_error, decl_event, decl_module, decl_storage, ensure, Parameter};
+use frame_system::ensure_root;
+use sp_arithmetic::traits::Saturating;
+use sp_arithmetic::traits::Zero;
+use sp_std::vec::Vec;
+
+use common::council::CouncilBudgetManager;
+use common::origin::MemberOriginValidator;
+use common::MemberId;
+
+use codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+/// Main pallet-bounty trait.
+pub trait Trait: frame_system::Trait + balances::Trait + common::Trait {
+    /// Events
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+    /// Bounty Id type
+    type BountyId: From<u32> + Parameter + Default + Copy;
+
+    /// Validates member ID and origin combination.
+    type MemberOriginValidator: MemberOriginValidator<Self::Origin, MemberId<Self>, Self::AccountId>;
+
+    /// Weight information for extrinsics in this pallet.
+    type WeightInfo: WeightInfo;
+
+    /// Provides an access for the council budget.
+    type CouncilBudgetManager: CouncilBudgetManager<BalanceOf<Self>>;
+}
+
+/// Alias type for the BountyParameters.
+pub type BountyCreationParameters<T> = BountyParameters<
+    BalanceOf<T>,
+    <T as frame_system::Trait>::BlockNumber,
+    <T as common::Trait>::MemberId,
+>;
+
+/// Defines who will be the oracle of the work submissions.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Copy, Debug)]
+pub enum OracleType<MemberId> {
+    /// Specific member will be the oracle.
+    Member(MemberId),
+
+    /// Council will become an oracle.
+    Council,
+}
+
+impl<MemberId> Default for OracleType<MemberId> {
+    fn default() -> Self {
+        OracleType::Council
+    }
+}
+
+/// Defines who can submit the work.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum AssuranceContractType<MemberId> {
+    /// Anyone can submit the work.
+    Open,
+
+    /// Only specific members can submit the work.
+    Closed(Vec<MemberId>),
+}
+
+impl<MemberId> Default for AssuranceContractType<MemberId> {
+    fn default() -> Self {
+        AssuranceContractType::Open
+    }
+}
+
+/// Defines parameters for the bounty creation.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct BountyParameters<Balance, BlockNumber, MemberId> {
+    /// Origin that will select winner(s), is either a given member or a council.
+    pub oracle: OracleType<MemberId>,
+
+    /// Contract type defines who can submit the work.
+    pub contract_type: AssuranceContractType<MemberId>,
+
+    /// Bounty creator member ID, should be None if created by a council.
+    pub creator_member_id: Option<MemberId>,
```

Comment:
Can we make this slightly more explicit so something like using a type

```Rust
enum BountyCreator {
Council,
Member(MemberId)
}
```

Reply:
Done.